sdk/eventhub/azure-eventhubs/azure/eventhub/_eventprocessor/ownership_manager.py

```diff
@@ -0,0 +1,142 @@
+        claimable_partition_ids = not_owned_partition_ids + timed_out_partition_ids
+
+        active_ownership = [ownership for ownership in ownership_list
+                            if ownership["last_modified_time"] + self.ownership_timeout >= now]
+        active_ownership_by_owner = defaultdict(list)
+        for ownership in active_ownership:
+            active_ownership_by_owner[ownership["owner_id"]].append(ownership)
+        active_ownership_self = active_ownership_by_owner[self.owner_id]
+
+        # calculate expected count per owner
+        all_partition_count = len(all_partition_ids)
+        # owners_count is the number of active owners. If self.owner_id is not yet among the active owners,
+        # then plus 1 to include self. This will make owners_count >= 1.
+        owners_count = len(active_ownership_by_owner) + \
+                       (0 if self.owner_id in active_ownership_by_owner else 1)
+        expected_count_per_owner = all_partition_count // owners_count
+        # end of calculating expected count per owner
+
+        to_claim = active_ownership_self
+        if len(active_ownership_self) < expected_count_per_owner:
+            # Either claims an inactive partition, or steals from other owners
+            if claimable_partition_ids:  # claim an inactive partition if there is
+                random_partition_id = random.choice(claimable_partition_ids)
+                random_chosen_to_claim = ownership_dict.get(
+                    random_partition_id,
+                    {"fully_qualified_namespace": self.fully_qualified_namespace,
+                     "partition_id": random_partition_id,
+                     "eventhub_name": self.eventhub_name,
+                     "consumer_group_name": self.consumer_group_name,
+                     }
+                )
+                random_chosen_to_claim["owner_id"] = self.owner_id
+                to_claim.append(random_chosen_to_claim)
+            else:  # steal from another owner that has the most count
+                active_ownership_count_group_by_owner = Counter(
+                    dict((x, len(y)) for x, y in active_ownership_by_owner.items()))
+                most_frequent_owner_id = active_ownership_count_group_by_owner.most_common(1)[0][0]
+                # randomly choose a partition to steal from the most_frequent_owner
+                to_steal_partition = random.choice(active_ownership_by_owner[most_frequent_owner_id])
+                to_steal_partition["owner_id"] = self.owner_id
+                to_claim.append(to_steal_partition)
+        return to_claim
+
+    def get_checkpoints(self):
+        if self.partition_manager:
+            checkpoints = self.partition_manager.list_checkpoints(
+                self.fully_qualified_namespace, self.eventhub_name, self.consumer_group_name)
+            return {x["partition_id"]: x for x in checkpoints}
```

Comment:
In this case it seems we're not actually getting checkpoints, rather we're getting the Partition IDs that have existing checkpoints?

Reply:
it's generating a dictionary, with key being `x["partition_id"]`, and value being `x`
`x` is a checkpoint object.